### DISCUSSION TITLE
[syntax] u64(a), u64:8(a), u64:(U0_argv, I_2, 8)

### DISCUSSION THREAD

**Discussion by iacore** [vnmakarov/mir#379] [Q&A] [answered]
What are those?

https://github.com/vnmakarov/mir/blob/928e28fb3acaa50d051a906e76a55cc48a556574/mir-tests/test15.mir#L6

`u64:(U0_argv, I_2, 8)` is in the code below

```
M0:	module
proto0:	proto	i32, u64:U0___s
	import	puts
main:	func	i32, i32:i0_argc, u64:U0_argv
	local	i64:i1_i, i64:i_0, i64:i_1, i64:I_2, i64:i_3, i64:i_4, i64:i_5
# 2 args, 7 locals
	mov	i1_i, 0
	bges	L3, i1_i, i0_argc
L1:
	ext32	I_2, i1_i
	call	proto0, puts, i_1, u64:(U0_argv, I_2, 8)
L2:
	mov	i_3, i1_i
	adds	i_3, i_3, 1
	mov	i1_i, i_3
	blts	L1, i1_i, i0_argc
L3:
	ret	0
	endfunc
	export	main
	endmodule

```

**Comment by vnmakarov** [comment]
> `u64:(U0_argv, I_2, 8)` is 

uint64 value in memory whose address is `U0_argv + I_2 * 8`.  It is an address where base and index regs are used (the constant displacement is zero).  The uint64 value is passed as the first arg of `puts` call.  It seems the code is a modified version of mir-code generated by `c2m`.

**Reply by iacore** [reply]
I don't see this feature documented. Is this the preferred way to do memory access in MIR?

**Reply by vnmakarov** [reply, marked as answer]
It is documented:

https://github.com/vnmakarov/mir/blob/8f04165320100018e91fe0944c24f3d76b504d51/MIR.md?plain=1#L201-L207
